### ISSUE TITLE
CookieAuthentication creates redirect loop if you automatically redirect an authenticated user to the Login page to use a different account

### ISSUE THREAD

**Issue by nillkitty** [dotnet/aspnetcore#23076]
### Describe the bug
I have a use case in which an authenticated user needs to switch accounts and return back to the page they were on.  If they fail this authentication I want them to stay logged in as their currently authenticated account, so I am **not** logging them out first.

If I redirect them to the page defined as `LoginPath` and I use the `ReturnUrl` querystring parameter to determine where they should come back to,  ASP.NET takes it upon itself it redirect them back immediately upon hitting the login page even though I did not write this redirect nor did they ever see the page or re-authenticate.  My redirect in combination with this phantom redirect creates a redirect loop and forces me to use a second querystring parameter that is different from the one defined in the CookieAuthenticationOptions just to work around this.

I would prefer if I could disable this logic so that a user can to sign in as many times as they want.

### To Reproduce
Redirect an authenticated user to to the page defined as the LogInPath with a query parameter (of the name defined in ReturnUrlParameter) to the page they're currently on.   They will be redirected back even if your login page has no code.    

### Exceptions (if any)
None

### Further technical details
- ASP.NET Core version:  3.1
- Include the output of `dotnet --info`
```
.NET Core SDK (reflecting any global.json):
 Version:   3.1.100
 Commit:    cd82f021f4

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18362
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.100\

Host (useful for support):
  Version: 3.1.0
  Commit:  65f04fb6db

.NET Core SDKs installed:
  2.1.701 [C:\Program Files\dotnet\sdk]
  3.1.100-preview1-014459 [C:\Program Files\dotnet\sdk]
  3.1.100 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0-preview1.19508.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0-preview1.19506.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.0-preview1.19506.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```
- The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version:  VS 2019 Enterprise

**Comment by blowdart** [comment]
This is outside the scope of the templates provided, you're going to have to scaffold the code out and write your own logic inside the login page.

**Comment by nillkitty** [comment]
> This is outside the scope of the templates provided, you're going to have to scaffold the code out and write your own logic inside the login page.

I'm not using any templates.  I've written my own authentication scheme.  Im not using Identity or any of the Identity UI.  The only thing I'm using that's stock is services.AddAuthentication().AddCookie() from ASP.Security.  If you're saying I need to roll my own security because I want users to be able to switch accounts, that's fine but it's certainly not documented or expected.

**Comment by blowdart** [comment]
Are you calling Signout before you Signin again?

**Comment by nillkitty** [comment]
> Are you calling Signout before you Signin again?

If you're asking if I'm calling Signout before sending them to the login page -- no -- if they fail to authenticate I want them to stay on the account they're signed in as.   If you're literally asking if I'm calling Signout before signing them in as the new account -- they never get that far.   ASP's authorization filter redirects them back to their ReturnUrl before they even *see* the login page.

> Documenting things we don't specifically support would be an almost infinite amount of docs :D

I'm not asking for the documentation of things that aren't supported.  I'm asking for the fact that adding Cookie Authentication to your project does things beyond authorizing a cookie to be documented.   

* I think it's reasonable to expect that the authorization filter will redirect someone to your login page if they haven't authenticated.  
* I think it's reasonable to expect that the authorization filter will redirect someone to your access denied page if they fail authorization of a page/controller.
* I don't think it's reasonable to expect that once someone authenticates they're prevented from viewing your login page again until they log out.

**Comment by blowdart** [comment]
Sure, but (you knew there was going to be a but), the templates are a starting point, somewhere to get up and running, it's the simple use cases.

The authorization filters do not, and will not ever, redirect users anywhere. What happens when authorization fails, either due to a user being anonymous or a user being logged in is quite specific to the authentication type used, and it's the authentication handlers that react to a failed authorization (be it 401 or 403). We did this by design because, for example, JWT authentication can't do redirections, the user agent may not be a browser at all.

Everything you want to do is possible, but we don't provide that code by default, because it's a rare use case, so you have to do the heavy lifting yourself.

**Comment by nillkitty** [comment]
> The authorization filters do not, and will not ever, redirect users anywhere. 
Then why is it doing so?  I'm not using any templates.  How do I prevent the redirect I'm seeing?

**Comment by blowdart** [comment]
You are using cookie authentication. It is doing the redirection in response to a failed authorization and you can't stop it, not unless you just want to return 401 or 403s.

**Comment by nillkitty** [comment]
1. The page that's doing the phantom redirect (the log in page) doesn't have any authorization requirements.

2. I created a project to showcase this issue without any other unrelated code,  and was able to reproduce the problem,  but in doing so I found the root cause and was able to resolve it.  The issue was caused by the filter we're using to authenticate guests automatically (to inject IP/network/AD-site based claims for guest users) signing the user in unnecessarily on the request to the login page,  so I simply made it ignore requests to the login page so that a SignInAsync does not already occur on the same HttpContext that requests the log in page (LogInPath).

I'm closing this case since I can work around the phantom redirect without using a second "return url" parameter name, however I still stand by the facts that something was redirecting *authenticated* users from the log in page to the ReturnUrl destination completely outside of my code.
